message_ix/tests/test_feature_vintage_and_active_years.py

```diff
@@ -0,0 +1,344 @@
+    scen.add_horizon(year=years, firstmodelyear=2010)
+    obs = scen.vintage_and_active_years()
+    exp = pd.DataFrame(
+        {
+            "year_vtg": (2000, 2000, 2010, 2010, 2020),
+            "year_act": (2010, 2020, 2010, 2020, 2020),
+        }
+    )
+    pdt.assert_frame_equal(exp, obs, check_like=True)  # ignore col order
+
+    # Add a technology, its lifetime, and period durations
+    scen.add_set("node", "foo")
+    scen.add_set("technology", "bar")
+    scen.add_par(
+        "duration_period", pd.DataFrame({"unit": "???", "value": 10, "year": years})
+    )
+    scen.add_par(
+        "technical_lifetime",
+        pd.DataFrame(
+            {
+                "node_loc": "foo",
+                "technology": "bar",
+                "unit": "???",
+                "value": 20,
+                "year_vtg": years,
+            }
+        ),
+    )
+
+    # part is before horizon
+    obs = scen.vintage_and_active_years(ya_args=("foo", "bar", "2000"))
+    exp = pd.DataFrame({"year_vtg": (2000,), "year_act": (2010,)})
+    pdt.assert_frame_equal(exp, obs, check_like=True)  # ignore col order
+
+    obs = scen.vintage_and_active_years(
+        ya_args=("foo", "bar", "2000"), in_horizon=False
+    )
+    exp = pd.DataFrame({"year_vtg": (2000, 2000), "year_act": (2000, 2010)})
+    pdt.assert_frame_equal(exp, obs, check_like=True)  # ignore col order
+
+    # fully in horizon
+    obs = scen.vintage_and_active_years(ya_args=("foo", "bar", "2010"))
+    exp = pd.DataFrame({"year_vtg": (2010, 2010), "year_act": (2010, 2020)})
+    pdt.assert_frame_equal(exp, obs, check_like=True)  # ignore col order
+
+    # part after horizon
+    obs = scen.vintage_and_active_years(ya_args=("foo", "bar", "2020"))
+    exp = pd.DataFrame({"year_vtg": (2020,), "year_act": (2020,)})
+    pdt.assert_frame_equal(exp, obs, check_like=True)  # ignore col order
+
+    # Advance the first model year
+    scen.add_cat("year", "firstmodelyear", years[-1], is_unique=True)
+
+    # Empty data frame: only 2000 and 2010 valid year_act for this node/tec;
+    # but both are before the first model year
+    obs = scen.vintage_and_active_years(
+        ya_args=("foo", "bar", years[0]), in_horizon=True
+    )
+    pdt.assert_frame_equal(
+        pd.DataFrame(columns=["year_vtg", "year_act"]), obs, check_dtype=False
+    )
+
+    # Exception is raised for incorrect arguments
+    with pytest.raises(
+        ValueError, match="At least 2 arguments are required if using `ya_args`"
+    ):
+        scen.vintage_and_active_years(ya_args=("foo"))
+
+
+def test_vintage_and_active_years2(test_mp):
+    scen = Scenario(test_mp, **SCENARIO["dantzig"], version="new")
+
+    # Add years of differing time length
+    years = [2000, 2005, 2010, 2015, 2020, 2030]
+    scen.add_horizon(year=years, firstmodelyear=2010)
+
+    # Check if default function call is valid
+    obs = scen.vintage_and_active_years()
+    exp = pd.DataFrame(
+        {
+            "year_vtg": (
```

Comment:
Because these take up many lines, and similar values occur repeatedly (perhaps with small differences that are hard to spot), it makes sense to e.g.:
- Use a global variable. This way it is clear when the same values are used in multiple tests.
- Write a short utility function to generate the values.

This can make the tests easier to read.

I will do these things as part of checking/tidying the modifications to the core code.